nova/hypervisor.go

```diff
@@ -0,0 +1,251 @@
+		JSONBody: n.body,
+		OkCodes:  []int{200, 204},
+	})
+	if err != nil {
+		return fmt.Errorf("Cannot change node state. Recieved code: %s.\nError: %v", resp.StatusCode, err)
+	}
+	logger.Info.Println("Node disabled.")
+	n.Enabled = false
+
+	return
+}
+
+// Enable change node state to enable
+func (n *Hypervisor) Enable() error {
+	url := n.client.ServiceURL("os-services", "enable")
+	resp, err := n.client.Request("PUT", url, gophercloud.RequestOpts{
+		JSONBody: n.body,
+		OkCodes:  []int{200, 204},
+	})
+	if err != nil {
+		logger.Error.Println("Cannot change node state.")
+		return fmt.Errorf("Recieved code: %s.\nError: %v", resp.StatusCode, err)
+	}
+	logger.Info.Println("Node enabled.")
+	n.Enabled = true
+
+	return nil
+}
+
+func (n *Hypervisor) isMigrated(vmID string, hostID string) (bool, error) {
+	vm := new(NovaServer)
+	url := n.client.ServiceURL("servers", vmID)
+	resp, err := n.client.Request("GET", url, gophercloud.RequestOpts{
+		OkCodes:  []int{200, 204},
+	})
+	if err != nil {
+		return false, fmt.Errorf("Cannot gather server %v information: %v", vmID, err)
+	}
+
+	if err = getJson(resp.Body, &vm); err != nil {
+		return false, fmt.Errorf("Cannot decode JSON: %v", err)
+	}
+
+	if vm.Server.HostID != hostID {
+		return true, nil
+	}
+
+	return false, nil
+}
+
+// MigrateVMs live migrate all VMs out of node
+func (h *Hypervisor) MigrateVMs() (err error) {
+	var wg sync.WaitGroup
+	if err = h.updateVMList(); err != nil {
+		return fmt.Errorf("Cannot update server list: ", err)
+	}
+
+	for _, vm := range *h.vms {
+		wg.Add(1)
+
+		go func(vmID string, hostID string) {
+			migrated := true
+			defer wg.Done()
+			for a := 1; a < liveMigrationRetry + 1; a++ {
+				er := adminactions.LiveMigrate(h.client, vmID, adminactions.LiveMigrateOpts{
+					BlockMigration: true,
+				})
+				if er.Result.Err == nil {
+					logger.Info.Printf("Attempt: %d. Request to migrate VM %s accepted\n", a, vmID)
+					migrated = false
+					break
+				}
+				logger.Warning.Printf("Attempt: %d. Cannot run migratation of VM %s: %v.\n", a, vmID, er.Result.Err)
+				time.Sleep(time.Duration(a * 10) * time.Second)
+			}
+			for counter := 0; !migrated ; counter++ {
+				migrated, err = h.isMigrated(vmID, hostID)
+				if err != nil {
+					logger.Warning.Printf("Cannot update VM: %v status.", vmID)
+				}
+				if migrated {
+					logger.Info.Printf("VM: %v has been migrated.", vmID)
+				} else {
+					logger.Info.Printf("VM: %v has not been migrated.", vmID)
+					time.Sleep(time.Duration(counter * 10) * time.Second)
+				}
+			}
+			if !migrated {
+				logger.Info.Printf("Cannot migrate VM: %v.", vmID)
+			}
+		}(vm.ID, vm.HostID)
+	}
+
+	if waitTimeout(&wg, h.timeOut) {
+		logger.Warning.Println("Time out waiting for live-migration.")
+	} else {
+		logger.Warning.Println("All VMs migrated")
+	}
+
+	return
+}
+
+func (n *Hypervisor) updateVMList() (err error) {
+	pager := servers.List(n.client, servers.ListOpts{
```

Comment:
Function returns pager with error inside. So it gonna fail during converting pager to servers list.